README.rst
arbitrator
==========

Negotiates and handles Content-Type / Accept header negotiation for you.

|Version| |Downloads| |Status| |License| |Coverage|

Overview
---------

Too often we roll our own serialization middleware based on Content-Type
or Accept HTTP headers, and too often it's done incorrectly.  This library
aims to provide a very simple interface for handling this task for you
so you can spend your time elsewhere.

Installing
----------

.. code:: bash

    pip install arbitrator

Examples
--------

.. code:: python

    >>> import json
    >>> import msgpack
    >>> from tornado import web
    >>> class MyHandler(arbitrator.MediaTypeMixn, web.RequestHandler):
    ...
    ...     def initalize(self):
    ...         media_types = {
    ...             'application/json': json,
    ...             'application/msgpack': msgpack,
    ...             'application/x-msgpack': msgpack
    ...         }
    ...         super(MyHandler, self).initalize(supported_media_types=media_types)
    ...
    ...     def get(self, account_id):
    ...         payload = self.load_payload(
    ...             self.request.headers['Content-Type'],
    ...             self.request.content,
    ...         )
    ...         # ... do stuff with the payload dictionary
    ...         content_type, stringified = self.dump_payload(
    ...             self.request.headers['Accept'],
    ...             payload,
    ...         )
    ...         self.set_status(200)
    ...         self.set_header('Content-Type', content_type)
    ...         self.finish(stringified)


Contributing
------------

If you want to help make this project better you are officially an awesome
person.

Pull requests or Github issues are always welcome.  If you want to contribute
a patch please do the following.

1. Fork this repo and create a new branch
2. Do work
3. Add tests for your work (Mandatory)
4. Submit a pull request
5. Wait for Coveralls and Travis-CI to run through your PR
6. I'll review it and merge it

As a note, code without sufficient tests will not be merged.

General Info
------------

+---------------+-------------------------------------------------+
| Source        | https://github.com/djt5019/arbitrator           |
+---------------+-------------------------------------------------+
| Status        | https://travis-ci.org/djt5019/arbitrator        |
+---------------+-------------------------------------------------+
| Download      | https://pypi.python.org/pypi/arbitrator         |
+---------------+-------------------------------------------------+
| Documentation | http://arbitrator.readthedocs.org/en/latest     |
+---------------+-------------------------------------------------+
| Issues        | https://github.com/djt5019/arbitrator           |
+---------------+-------------------------------------------------+

.. |Version| image:: https://pypip.in/version/arbitrator/badge.svg
   :target: https://pypi.python.org/pypi/arbitrator

.. |Downloads| image:: https://pypip.in/d/arbitrator/badge.svg
   :target: https://pypi.python.org/pypi/arbitrator

.. |Status| image:: https://travis-ci.org/djt5019/arbitrator.svg
   :target: https://travis-ci.org/djt5019/arbitrator

.. |License| image:: https://pypip.in/license/arbitrator/badge.svg
   :target: https://arbitrator.readthedocs.org/

.. |Coverage| image:: https://coveralls.io/repos/djt5019/arbitrator/badge.svg?branch=master
   :target: https://coveralls.io/r/djt5019/arbitrator?branch=master


docs/history.rst
.. :changelog:

Release History
===============

Next Release
------------

* Initial release the library
* Add Tornado mix-in for Content-Type / Accept header negotiation


web-example.py
import datetime
import json
import logging

from arbitrator.tornado import ArbitratedApplication, ArbitratedHandler
from tornado import ioloop, web
import yaml


class YamlHandler(object):
    @staticmethod
    def dumps(data):
        return yaml.safe_dump(data, default_flow_style=False)
    @staticmethod
    def loads(data_str):
        return yaml.safe_load(data_str)


class JsonHandler(object):
    @staticmethod
    def dumps(data):
        return json.dumps(data).encode('utf8')
    @staticmethod
    def loads(data_str):
        return json.loads(data_str.decode('utf8'))


class DateHandler(ArbitratedHandler, web.RequestHandler):
    def get(self, tz):
        if tz == 'utc':
            now = datetime.datetime.utcnow()
        elif tz == 'local':
            now = datetime.datetime.now()
        else:
            raise web.HTTPError(400)

        rep = {
            'month': now.month,
            'day': now.day,
            'year': now.year,
        }
        self.write(self.encode_response(rep))


class Adder(ArbitratedHandler, web.RequestHandler):
    def post(self):
        data = self.decode_request()
        result = sum(float(x) for x in data['addends'])
        self.write(self.encode_response({'result': result}))


def make_app():
    return ArbitratedApplication(
        handlers=[
            web.url(r'/now/(utc|local)', DateHandler),
            web.url(r'/adder', Adder),
        ],
        content_handlers=[
            ('application/json', JsonHandler),
            ('application/yaml', YamlHandler),
        ],
        debug=True,
    )


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.DEBUG,
        format='%(levelname)1.1s - %(name)s: %(message)s.')
    app = make_app()
    app.listen(8000)
    ioloop.IOLoop.current().start()


arbitrator/tornado.py
import logging

from ietfparse import algorithms, errors, headers
from tornado import web


_logger = logging.getLogger(__name__)


class ArbitratedApplication(web.Application):

    def __init__(self, *args, **kwargs):
        self._default_content_type = kwargs.pop('default_content_type',
                                                'application/octet-stream')
        self._supported_mime_types = set()
        self._handlers = {}
        for content_type, handler in kwargs.pop('content_handlers', []):
            parsed = headers.parse_content_type(content_type)
            self._supported_mime_types.add(parsed)
            self._handlers[parsed] = handler

        super(ArbitratedApplication, self).__init__(*args, **kwargs)

    def decode_request(self, request):
        content_type = headers.parse_http_accept_header(
            request.headers.get('Content-Type', self._default_content_type))
        try:
            selected, _ = algorithms.select_content_type(
                content_type, self._supported_mime_types)
            handler = self._handlers[selected]
        except errors.NoMatch:
            raise web.HTTPError(415)
        return handler.loads(request.body)

    def encode_response(self, request, response_body):
        accept_target = request.headers.get(
            'Accept', self._default_content_type)
        _logger.debug('determining response type, '
                      'client wants %s and I can produce %s',
                      accept_target, self._supported_mime_types)
        acceptable = headers.parse_http_accept_header(accept_target)
        try:
            selected, _ = algorithms.select_content_type(
                acceptable, self._supported_mime_types)
        except errors.NoMatch:
            selected = self._default_content_type

        _logger.debug('selected content type %s', selected)
        try:
            handler = self._handlers[selected]
        except KeyError:
            raise web.HTTPError(406)
        _logger.debug('dumping with %r', handler)
        return str(selected), handler.dumps(response_body)


class ArbitratedHandler(object):

    def decode_request(self):
        return self.application.decode_request(self.request)

    def encode_response(self, response_body):
        content_type, body = self.application.encode_response(
            self.request, response_body)
        self.set_header('Content-Type', content_type)
        return body


arbitrator/__init__.py
"""
Arbitrator
==========

"""
version_info = (1, 0, 0)
__version__ = '.'.join(str(v) for v in version_info)


from ietfparse.algorithms import select_content_type
from ietfparse.errors import NoMatch
from ietfparse.headers import parse_content_type, parse_http_accept_header


class MediaTypeMixin(object):

    default_media_type = None
    """Media type to use when nothing else matches the Accept header."""

    media_types = None
    """Mapping between media types and their serializer."""

    supported = set()
    """Set of supported IETF Accept headers."""

    def initialize(self, *args, **kwargs):
        if MediaTypeMixin.media_types is None:
            types = kwargs.pop('media_types', {})
            supported = {parse_content_type(mime) for mime in types}
            MediaTypeMixin.media_types = media_types
            MediaTypeMixin.supported = supported

        if MediaTypeMixin.default_media_type is None:
            MediaTypeMixin.default_media_type = kwargs.pop(
                'default_media_type')

        self.__parsed_content_type = None
        self.__parsed_accept_headers = None

        super(MediaTypeMixin, self).initialize(*args, **kwargs)

    @property
    def content_type_header(self):
        """Returns the Content-Type header for the time being."""
        if self.__parsed_content_type is None:
            header = self.request.headers.get(
                'Content-Type', self.default_media_type)

            if header is None:
                raise web.HTTPError(415)

            self.__parsed_content_type = parse_content_type(header)

        return self.__parsed_content_type

    @property
    def accept_headers(self):
        """Returns all Accept headers."""
        if self.__parsed_accept_headers:
            header = self.request.headers.get(
                'Accept', self.default_media_type)

            if header is None:
                raise web.HTTPError(406)

            self.__parsed_accept_headers = parse_http_accept_header(header)

        return self.__parsed_accept_headers

    def load_payload(self, **kwargs):
        """Deserializes ``payload`` with the provided Content-Type header.

        If the request ``payload`` cannot be deserialized we will return
        a 400.  If there no support registered for the Content-Type header
        then we will raise a 415.

        :param dict payload:
            The Python dictionary containing the data.

        :param dict kwargs:
            Additional keyword args passed to the serializer.

        :returns: Dictionary containing the request data.

        """
        header = str(self.content_type_header)
        try:
            return self.media_types[header].loads(self.request.content)
        except ValueError:
            reason = 'Unable to load payload with {0} header'.format(header)
            raise web.HTTPError(400, reason=reason)

    def dump_payload(self, payload, **kwargs):
        """Serializes ``payload`` with the proper accept type.

        Returns a tuple consisting of the media type selected for
        serialization and the serialized data.  If the ``payload`` cannot
        be serialized we will return a 400.  If there no support registered
        for the accept header the we will raise a 406.

        :param dict payload:
            The dictionary containing the data to be serialized.

        :param dict kwargs:
            Additional keyword args passed to the serializer.

        :returns:
            Tuple containing the selected Accept header used and
            the serialized data matching the selected header.

        """
        accept = self.accept_headers

        try:
            requested, selected = select_content_type(self.supported, accept)
            selected = str(selected)
            serializer = self.media_types[selected].dumps
            return selected, serializer(payload, **kwargs)
        except NoMatch:
            raise web.HTTPError(406, )
        except ValueError:
            reason = 'Unable to dump payload with {0} header'.format(selected)
            raise web.HTTPError(400, reason=reason)
